Data protection template using Protegrity

ilya-kozyrev opened the pull request:
Some users may want to protect their sensitive data using tokenization.
We propose to create a template that will provide integration with protection RPC server using Beam transform to protect sensitive data using tokenization. 

At a high level, a template that will:

  *   support batch (GCS) and streaming (Pub/Sub) input sources
  *   tokenize sensitive data via external RPC service - we are about to use Protegrity
  *   output tokenized data into BigQuery, BigTable, or GCS
  *  supported formats CSV. JSON

More details and the proposed design are available in the design [doc](https://docs.google.com/document/d/1fnsUfGpCx8A_MBchPRvlm4gU0Ai5EQNSiZS1mg_A_zg/edit?usp=sharing).

KhaninArtur commented:
@googlebot I consent.

ramazan-yapparov commented:
@googlebot I consent.

daria-malkova commented:
@googlebot I consent

daria-malkova commented:
@googlebot I fixed it

AKosolapov commented:
@googlebot I consent.

MikhailMedvedevAkvelon commented:
@googlebot I fixed it..

MikhailMedvedevAkvelon commented:
@googlebot I consent.

Nuzhdina-Elena commented:
@googlebot I consent.

RaphaelSanamyan commented:
@googlebot I fixed it.

ilya-kozyrev commented:
Hi, @prathapreddy123  Did you have a chance to look at this PR?

We merged similar [PR](https://github.com/apache/beam/pull/13995) with the e2e example in the Beam repository.

prathapreddy123 commented:
> Hi, @prathapreddy123 Did you have a chance to look at this PR?
> 
> We merged similar [PR](https://github.com/apache/beam/pull/13995) with the e2e example in the Beam repository.

Hi @ilya-kozyrev - Not yet. Will check in the current week

prathapreddy123 reviewed (changes requested):
Thanks for Contributing.  I believe several classes (if carefully designed) can be reused by promoting to common module across the templates in future.  Included few ideas for reference.

ilya-kozyrev commented:
> Thanks for Contributing. I believe several classes (if carefully designed) can be reused by promoting to common module across the templates in future. Included few ideas for reference.

Thank you for your review. That's an awesome idea to make io classes more generic and promote them to the common module. I think it will be a good next step. 

To avoid increasing PR size and reduce the effort to review, maybe make sense to use internal io classes in this template as is? 
After this PR will be merged, we'll create separate PRs for each transform to promote them into the common module based on your suggestions. 
When all transforms will be implemented and merged, we will create another PR to refactor the current template.

What do you think?

prathapreddy123 commented:
> > Thanks for Contributing. I believe several classes (if carefully designed) can be reused by promoting to common module across the templates in future. Included few ideas for reference.
> 
> Thank you for your review. That's an awesome idea to make io classes more generic and promote them to the common module. I think it will be a good next step.
> 
> To avoid increasing PR size and reduce the effort to review, maybe make sense to use internal io classes in this template as is?
> After this PR will be merged, we'll create separate PRs for each transform to promote them into the common module based on your suggestions.
> When all transforms will be implemented and merged, we will create another PR to refactor the current template.
> 
> What do you think?

Sure we can consider that approach. But considering this pipeline has already been contributed to Beam repo, instead of replicating the same we can demonstrate patterns through building reusable classes and reduce effort considerably for new templates.

ilya-kozyrev commented:
@googlebot I consent.

RaphaelSanamyan commented:
@googlebot I consent.

ilya-kozyrev commented:
@prathapreddy123 We've refactored modules and classes to common that you suggest and changed our code/docs based on your suggestions. Could you please review the changes?

ilya-kozyrev commented:
Hi @prathapreddy12. Did you have a chance to review our changes?

prathapreddy123 commented:
> Hi @prathapreddy12. Did you have a chance to review our changes?

Hi @ilya-kozyrev -  We were currently undergoing through design & refactoring process. It would be helpful to review the template post completion of our design to share latest updates. Expected ETA  2 - 4 weeks.

ilya-kozyrev commented:
Hi @prathapreddy123 could you please review this PR?  If any help from our side is required please let us know. 
Thanks in advance

prathapreddy123 commented:
> Hi @prathapreddy123 could you please review this PR? If any help from our side is required please let us know.
> Thanks in advance

@an2x -  Can you take care of this request. For context, please refer to earlier comments

an2x commented:
> > Hi @prathapreddy123 could you please review this PR? If any help from our side is required please let us know.
> > Thanks in advance
> 
> @an2x - Can you take care of this request. For context, please refer to earlier comments

Will do, probably early next week.

ilya-kozyrev commented:
Hi @an2x, @prathapreddy123. Did you have a chance to review this PR? Can we help with it somehow?

ilya-kozyrev commented:
Hi @prathapreddy123 @an2x. We are looking forward to your reviewing. We believe that our improvements could be helpful for the community. If we can help somehow with merging this PR, please free to ask.

bvolpato commented:
Hi @ilya-kozyrev -- thanks for your contribution here. Unfortunately we ran behind on reviews, and the structure of the project changed a bit. Is there still interest in pursuing this template? We can work together on a plan to resolve conflicts + review the code.

bvolpato commented:
Closing this because it got stale for a few months. Please rebase / file a new one to bring it up to the top of the stack.
